runtime/compiler/z/codegen/J9TreeEvaluator.cpp

```diff
@@ -10762,10 +10748,14 @@ genInitArrayHeader(TR::Node * node, TR::Instruction *& iCursor, bool isVariableL
 
    static char * allocZeroArrayWithVM = feGetEnv("TR_VMALLOCZEROARRAY");
 
-   // Zero size arrays are discontiguous. If the array is not discontiguous, it writes the length in first element which should not cause any issue.
    if(node->canSkipZeroInitialization() && allocZeroArrayWithVM == NULL)
+      {
+      // Zero size arrays are discontiguous. If the array is not discontiguous, it writes the length in first element which should not cause any issue.
       iCursor = generateRXInstruction(cg, TR::InstOpCode::ST, node, eNumReg,
                       generateS390MemoryReference(resReg, fej9->getOffsetOfDiscontiguousArraySizeField(), cg), iCursor);
+      // Clear padding in contiguous array header layout. +4 because size field is 4 bytes wide.
+      iCursor = generateRXInstruction(cg, TR::InstOpCode::ST, node, eNumReg, generateS390MemoryReference(resReg, static_cast<int32_t>(fej9->getOffsetOfDiscontiguousArraySizeField()) + 4, cg), iCursor);
```

Comment:
Can you explain the comment ?

Reply:
Extra padding is added to the array header for 8 byte boundary alignment; so with that comment I just meant that we add 4 to size field offset to get to the padding. I'll update the comment to be more clear.

Reply:
Okk, does `enumReg`  contains 0 ?

Reply:
So this works because if `enumReg` isn't 0 (i.e. we aren't dealing with 0 size array), whatever we write at `getOffsetOfDiscontiguousArraySizeField` will be overwritten by the element or dataAddr field (if present). But I think using `XC` might provide better readability. I'll update the code.